habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -386,6 +386,286 @@ def step(self, *args, is_last_action, **kwargs):
+            v_raw = self._velocity_feedback_control(
+                np.sign(xyt_err[0])*lin_err_abs, self.acc_lin, self.v_max
+            )
+            v_limit = self._turn_rate_limit(
+                lin_err_abs,
+                heading_err_abs,
+                self.w_max / 2.0,
+                tol=self.lin_error_tol,
+            )
+            # Allow the agent to move backward
+            v_cmd = np.clip(v_raw, -v_limit, v_limit)
+
+            # Compute angular velocity
+            w_cmd = self._velocity_feedback_control(
+                heading_err, self.acc_ang, self.w_max
+            )
+
+        # Rotate to correct yaw if yaw tracking is on and XY position is at goal
+        elif abs(ang_err) > self.ang_error_tol and self.track_yaw:
+            # Compute angular velocity
+            w_cmd = self._velocity_feedback_control(
+                ang_err, self.acc_ang, self.w_max
+            )
+
+        # v_cmd and w_cmd will be clip for [-1, 1]
+        return v_cmd, w_cmd
+
+    def _transform_global_to_base(self, XYT, current_pose, sim=None):
+        """
+        Transforms the point cloud into geocentric frame to account for
+        camera position
+        Input:
+            XYT                     : target goal ...x3
+            current_pose            : base position (x, y, theta (radians))
+        Output:
+            XYT : ...x3
+        """
+
+        trans = sim.robot.base_transformation
+        local_pos = trans.inverted().transform_point(sim.robot.base_pos)
+        goal_pos = trans.inverted().transform_point(
+            np.array([XYT[0], sim.robot.base_pos[1], XYT[1]])
+        )
+
+        error_t = XYT[2] - current_pose[2]
+        error_x = goal_pos[0] - local_pos[0]
+        error_y = goal_pos[1] - local_pos[1]
+
+        return [error_x, error_y, error_t]
+
+
+    def forward(self, xyt, sim):
+        xyt_err = self._compute_error_pose(xyt, sim)
+        return self._feedback_simple(xyt_err)
+
+
+@registry.register_task_action
+class BaseWaypointVelAction(RobotAction):
+    """
+    The robot base motion is constrained to the NavMesh and controlled with velocity commands integrated with the VelocityControl interface.
+    Optionally cull states with active collisions if config parameter `allow_dyn_slide` is True
+    The robot is given a target waypoint and output a velocity command
+    """
+
+    def __init__(self, *args, config, sim: RearrangeSim, **kwargs):
+        super().__init__(*args, config=config, sim=sim, **kwargs)
+        self._sim: RearrangeSim = sim
+        self.base_vel_ctrl = habitat_sim.physics.VelocityControl()
+        self.base_vel_ctrl.controlling_lin_vel = True
+        self.base_vel_ctrl.lin_vel_is_local = True
+        self.base_vel_ctrl.controlling_ang_vel = True
+        self.base_vel_ctrl.ang_vel_is_local = True
+        self.pid_controller = Controller(self._config.lin_speed, self._config.ang_speed)
+
+    @property
+    def action_space(self):
+        lim = 20
```

Comment:
Where is this number coming from?

Reply:
I am not sure actually, the original velocity controller also uses this